Src/WindowsServer/WindowsServer.Shared/Implementation/AzureHeartbeatProperties.cs

```diff
@@ -0,0 +1,95 @@
+
+    internal class AzureHeartbeatProperties
+    {
+        // internal for testing
+        internal readonly List<string> DefaultFields = new List<string>()
+        {
+            "osType",
+            "location",
+            "name",
+            "offer",
+            "platformFaultDomain",
+            "platformUpdateDomain",
+            "publisher",
+            "sku",
+            "version",
+            "vmId",
+            "vmSize",
+            "subscriptionId",
+            "resourceGroupName"
+        };
+
+        /// <summary>
+        /// Flags that will tell us whether or not Azure VM metadata has been attempted to be gathered or not, and
+        /// if we should even attempt to look for it in the first place. Static to ensure we only ever attempt this
+        /// once for any given consuming application.
+        /// </summary>
+        private static bool isAzureMetadataCheckCompleted = false;
+
+        private IAzureMetadataRequestor azureInstanceMetadataRequestor = null;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureHeartbeatProperties"/> class.
+        /// </summary>
+        public AzureHeartbeatProperties() : this(null, false)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureHeartbeatProperties"/> class.
+        /// </summary>
+        /// <param name="azureInstanceMetadataHandler">For testing: Azure metadata request handler to use when requesting data from azure specifically. If left as null, an instance of AzureMetadataRequestor is used.</param>
+        /// <param name="resetCheckCompleteFlag">For testing: set to true to reset the check that we've already acquired this data.</param>
+        internal AzureHeartbeatProperties(IAzureMetadataRequestor azureInstanceMetadataHandler = null, bool resetCheckCompleteFlag = false)
+        {
+            this.azureInstanceMetadataRequestor = azureInstanceMetadataHandler;
+            if (this.azureInstanceMetadataRequestor == null)
+            {
+                this.azureInstanceMetadataRequestor = new AzureMetadataRequestor();
+            }
+
+            if (resetCheckCompleteFlag)
+            {
+                AzureHeartbeatProperties.isAzureMetadataCheckCompleted = false;
+            }
+        }
+
+        public bool IsKeyword(string keyword)
+        {
+            return this.DefaultFields.Contains(keyword, StringComparer.OrdinalIgnoreCase);
+        }
+
+        public async Task<bool> SetDefaultPayload(IEnumerable<string> disabledFields, IHeartbeatPropertyManager provider)
+        {
+            bool hasSetFields = false;
+
+            if (!AzureHeartbeatProperties.isAzureMetadataCheckCompleted)
+            {
+                // only ever attempt this once
+                AzureHeartbeatProperties.isAzureMetadataCheckCompleted = true;
+
+                var allFields = await this.azureInstanceMetadataRequestor.GetAzureInstanceMetadataComputeFields()
```

Comment:
Do we consider metadata check completed even if this fails due to legitimate reason on IMS metadata side (e.g. some throttling at the moment we tried to access it)? If not - then we might need to move `isAzureMetadataCheckCompleted` assignment after this call.

Reply:
Yes, let's consider it failed at this point in the code. Any retries we decide to implement can be handled within the `IAzureMetadataRequestor` implementation, should we deem that important enough to do.